### ISSUE TITLE
Cannot login to heroku -- cannot unmarshal DNS message

### ISSUE THREAD

**Issue by kalyandechiraju** [heroku/toolbelt#165]
I am using heroku toolbelt on a Mac running 10.11.5 and when I do heroku login and enter my credentials, I get this error. This is very inconvenient and I cannot use heroku on my machine. 

`Post https://api.heroku.com/login: dial tcp: lookup api.heroku.com on 8.8.4.4:53: cannot unmarshal DNS message`

Tried many things like `sudo heroku login`, but didnt help. Please let me know if there is any workaround or fix.

> heroku version
> heroku-toolbelt/3.43.5 (universal.x86_64-darwin15) ruby/2.0.0
> heroku-cli/5.2.24-4b7e305 (darwin-amd64) go1.6.2
> You have no installed plugins.

**Comment by ransombriggs** [comment]
Could you paste in the output of the following command?  This is due to a bug with the programming library Go and DNS responses that are not to spec and are larger than a certain size.  I have not been been able to recreate this locally, but hopefully we can figure out the difference in our setups and track down why you are getting a too large response.  (I also run 10.11.5)
```
dig @8.8.4.4 api.heroku.com
```

**Comment by kalyandechiraju** [comment]
; <<>> DiG 9.8.3-P1 <<>> @8.8.4.4 api.heroku.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 55215
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 13, ADDITIONAL: 12

;; QUESTION SECTION:
;api.heroku.com.			IN	A

;; ANSWER SECTION:
api.heroku.com.		21	IN	CNAME	midgard.heroku.com.
midgard.heroku.com.	441	IN	CNAME	midgard.herokussl.com.
midgard.herokussl.com.	3137	IN	CNAME	elb082153-1559744486.us-east-1.elb.amazonaws.com.
elb082153-1559744486.us-east-1.elb.amazonaws.com. 51 IN	A 54.243.105.250
elb082153-1559744486.us-east-1.elb.amazonaws.com. 51 IN	A 54.225.188.133
elb082153-1559744486.us-east-1.elb.amazonaws.com. 51 IN	A 23.21.149.112

;; AUTHORITY SECTION:
com.			140079	IN	NS	g.gtld-servers.net.
com.			140079	IN	NS	j.gtld-servers.net.
com.			140079	IN	NS	i.gtld-servers.net.
com.			140079	IN	NS	b.gtld-servers.net.
com.			140079	IN	NS	k.gtld-servers.net.
com.			140079	IN	NS	a.gtld-servers.net.
com.			140079	IN	NS	e.gtld-servers.net.
com.			140079	IN	NS	c.gtld-servers.net.
com.			140079	IN	NS	l.gtld-servers.net.
com.			140079	IN	NS	m.gtld-servers.net.
com.			140079	IN	NS	f.gtld-servers.net.
com.			140079	IN	NS	h.gtld-servers.net.
com.			140079	IN	NS	d.gtld-servers.net.

;; ADDITIONAL SECTION:
g.gtld-servers.net.	124660	IN	A	192.42.93.30
j.gtld-servers.net.	91373	IN	A	192.48.79.30
i.gtld-servers.net.	119024	IN	A	192.43.172.30
b.gtld-servers.net.	119025	IN	A	192.33.14.30
k.gtld-servers.net.	27063	IN	A	192.52.178.30
a.gtld-servers.net.	16745	IN	A	192.5.6.30
e.gtld-servers.net.	123106	IN	A	192.12.94.30
c.gtld-servers.net.	153632	IN	A	192.26.92.30
l.gtld-servers.net.	22933	IN	A	192.41.162.30
f.gtld-servers.net.	107336	IN	A	192.35.51.30
h.gtld-servers.net.	29637	IN	A	192.54.112.30
d.gtld-servers.net.	119025	IN	A	192.31.80.30

;; Query time: 48 msec
;; SERVER: 8.8.4.4#53(8.8.4.4)
;; WHEN: Thu Jul 14 07:37:57 2016
;; MSG SIZE  rcvd: 609

**Comment by kalyandechiraju** [comment]
@ransombriggs Can you help me out?

**Comment by ransombriggs** [comment]
@kalyandechiraju sorry, but I am on vacation right now, @dickeyxxx can you see if you can reproduce?

**Comment by ransombriggs** [comment]
@dickeyxxx fwiw, last time someone saw this I could not reproduce even when using the same DNS server which made me think it was a configuration issue. That time we were on slightly different OS versions, but this time it is the same.  If the 'dig' command does immediately reproduce for you I would post it to stack exchange because I wasted too much time trying to figure it out last time. The fact that it is a public DNS server should make it easier for a DNS expert to help.

**Comment by ransombriggs** [comment]
@kalyandechiraju Sorry for the delayed response, back from vacation and looking into what I had tried before and came across this https://github.com/heroku/heroku/issues/1899 where I shipped a change that appeared to fix things for two users.  What is the output of `heroku version`?  I want to make sure that you have a version after the fix.

**Comment by kalyandechiraju** [comment]
@ransombriggs 
heroku-toolbelt/3.43.5 (universal.x86_64-darwin15) ruby/2.0.0
heroku-cli/5.2.24-4b7e305 (darwin-amd64) go1.6.2
You have no installed plugins.

**Comment by ransombriggs** [comment]
@kalyandechiraju could you run `heroku update` and paste in the output?

**Comment by kalyandechiraju** [comment]
@ransombriggs 

heroku update
 ▸    Get https://cli-assets.heroku.com/branches/stable/manifest.json: dial tcp: lookup cli-assets.heroku.com on 8.8.4.4:53: cannot unmarshal DNS message
 !    Use `brew upgrade heroku` to update

$ brew upgrade heroku
Error: heroku 3.43.5 already installed

**Comment by ransombriggs** [comment]
@kalyandechiraju could you do me a favor and remove `~/.local/share/heroku/` and run `heroku update` again?

**Comment by kalyandechiraju** [comment]
@ransombriggs I did that

heroku update
heroku-cli: Installing CLI... 21.48MB/21.48MB
 !    Use `brew upgrade heroku` to update

After download, it asked me to run brew upgrade heroku...
$ brew upgrade heroku
Error: heroku 3.43.5 already installed

Same thing repeated, still heroku login not working :(

**Comment by ransombriggs** [comment]
@kalyandechiraju what is the output of `heroku version` now?

**Comment by kalyandechiraju** [comment]
@ransombriggs  Hey something new happened. I tried to run install script inside local/share/heroku then the folder got cleared and when I ran heroku login, cli updated to new version

heroku login
heroku-cli: Updating to 5.2.34-65b16c9... 15.5 MB/15.5 MB
 ▸    fork/exec /usr/local/bin/heroku: resource temporarily unavailable

Now am facing this new error

heroku login
 ▸    fork/exec /usr/local/bin/heroku: resource temporarily unavailable

**Comment by kalyandechiraju** [comment]
$ heroku version
 ▸    fork/exec /usr/local/bin/heroku: resource temporarily unavailable

**Comment by ransombriggs** [comment]
@kalyandechiraju It looks like the brew installation is messed up, what would you think of uninstalling from brew and doing an install from `https://toolbelt.heroku.com/`?

**Comment by kalyandechiraju** [comment]
@ransombriggs  Yes. I removed brew heroku and installed a fresh copy from the url above.

heroku version
heroku-toolbelt/3.43.9 (x86_64-darwin10.8.0) ruby/1.9.3
heroku-cli/5.2.34-65b16c9 (darwin-amd64) go1.6.2
You have no installed plugins.

Still heroku login
` ▸    Post https://api.heroku.com/login: dial tcp: lookup api.heroku.com on 8.8.4.4:53: cannot unmarshal DNS message`

Is it just for me ? :(

**Comment by ransombriggs** [comment]
@kalyandechiraju could you try running the following?
```
GODEBUG=netdns=cgo heroku login
```

**Comment by kalyandechiraju** [comment]
@ransombriggs got same Post https://api.heroku.com/login: dial tcp: lookup api.heroku.com on 8.8.4.4:53: cannot unmarshal DNS message

**Comment by ransombriggs** [comment]
@kalyandechiraju Could you try `GODEBUG=netdns=go heroku login`?  I am assuming it will not work because I think it is the default behavior, but it was suggested here http://stackoverflow.com/questions/38770700/golang-issue-with-dns-response-512-bytes-on-mac-os-x-cannot-unmarshal-dns-mes/38777375 and I also am trying to figure out how to replicate your behavior here http://serverfault.com/questions/794872/mac-os-x-dns-issues-with-google-dns/794883#794883 but have not had any luck so far.

**Comment by ransombriggs** [comment]
@kalyandechiraju @dickeyxxx suggested that the following may work for you since it does not use the same login method as the code you are running.  Basically `heroku-accounts` is a plugin that people use to manage multiple heroku accounts, but in your case we are using to avoid hitting the problematic Go code.
```
heroku plugins:install heroku-accounts
heroku accounts:add <account_email>
```

**Comment by wlhaskell** [comment]
@ransombriggs I had some students experiencing this same issue and the workaround was successful.  Thank you for the help.
